### DISCUSSION TITLE
Can you use a join in a row level security policy?

### DISCUSSION THREAD

**Discussion by pranavbadami** [supabase/supabase#811] [Questions] [answered]
Hey folks, love the Supabase product! 

I'm currently trying to implement a join in a row level security policy and looking for pointers on how to go about this (slash if it's even possible). 

Here are the tables I'm working with:
- `auth.users` -- Supabase auth table
- `app_users` -- Public users table for app users
    - has a column `auth_users_id` that is a reference to `auth.users.uid` (side question: is it possible to foreign key here? I didn't find a way to via the console)
    - has a foreign key to the `teams` table (see next)
- `teams` -- a workspace that an `app_user` can belong to, an `app_user` can belong to multiple `teams`

The weird constraint here is that multiple `app_users`, each pointing to a different `team`, can point to the same `auth.users` row. So, for example, you could have this situation:

#### `auth.users`:
```
| uid        | ... |
|------------|-----|
| example123 | ... |
| example456 | ... |
```

#### `teams`:

```
| id | name  | ... |
|----|-------|-----|
| 1  | team1 | ... |
| 2  | team2 | ... |
```

#### `app_users`:
```
| id | auth_user_uid | team_id |
|----|---------------|---------|
| 1  | example123    | 1       | <--- 
| 2  | example123    | 2       | <---
| 3  | example456    | 2       |
```
Now, let's say I want to enable row level security on the `teams` table, such that a Supabase `auth.user` can only access rows for teams that their `app_users` belong to. So, in the above example, `example123` should be able to access `team1` and `team2`, while `example456` should only be able to access `team2`.

What would the row level policy for this look like? Is it possible to implement this with modifications to the schema that I'm missing? It's weird because I'd have to join onto the `app_users` table based on the `auth.uid()` value and then check if `team.id` is in a set of `team_id` values. Does that make sense? 

Any help would be greatly appreciated, thanks :pray:

**Comment by steve-chavez** [comment, marked as answer]
Hey @pranavbadami,

> Can you use a join in a row level policy?

In short: yes, you can.

I do this for this policy [here](https://github.com/steve-chavez/socnet/blob/master/security/users.sql#L82-L92). The policy calls a function that executes a [complex query with a join](https://github.com/steve-chavez/socnet/blob/master/schema/util.sql#L31-L57). Of course you can inline the query as well.

One thing to consider is that the policies apply recursively. So if you have a policy defined for the target join table, it will also kick in on the policy of the source join table.

I don't have the SQL for your particular use case now, but I hope the above helps finding your way.

(The above complex query on the policy is just an example, it's recommended to keep the policy query simple so it can be fast)

**Reply by pranavbadami** [reply]
Hey @steve-chavez ,

Thanks for the quick reply and pointing me in the right direction! 

After writing the above post, I actually came across [this Stack Overflow article](https://stackoverflow.com/questions/41354818/postgresql-row-level-security-involving-a-view-or-a-select-with-join) asking a very similar question. I was effectively able to map my schema above onto the SO answer's policy and it looks like this: 

```
EXISTS(
  SELECT
    teams.id,
    teams.name 
  FROM teams 
  INNER JOIN app_users ON (app_users.team = teams.id) 
  INNER JOIN auth.users AS auth_users ON (auth_users.id = app_users.auth_user_uid) 
  WHERE auth_users.id = auth.uid()
)
```

Then, for other tables in my schema where I have `team_id` as a foreign key, I just need to replace `teams` with the other table names in my policies. I think that should work!

Do you have any insight on the performance implications of this? My app is still pretty small but I do some have tables with tens of thousands of rows, so I'm hoping queries against those tables execute in a reasonable time.

**Reply by steve-chavez** [reply]
The policy looks good. You can add indexes for the join/where conditions to make sure the query is fast.

That being said, it's always best to check the perf of the query for yourself. For RLS you can do this:

```sql
begin;

set local role authenticated; -- or to other roles for which RLS should be applied

EXPLAIN ANALYZE SELECT * FROM app_users;
-- this would give you the exec/planning time and the strategy used by the pg planner.
-- The policy query will be applied here

commit;
```

**Reply by pranavbadami** [reply]
So I realized I was overthinking this and ended up solving this with a much simpler policy that doesn't need joins. All I needed to do was check if a row existed in the `app_users` table that had the same `team` value and `auth_user_uid` as the accessing user. 

The policy that worked for me (after some tinkering 😅) was: 
```
EXISTS (
   SELECT app_users.slack_id,
    app_users.name
   FROM app_users
  WHERE app_users.team = teams.slack_id 
  AND app_users.auth_users_uid = uid()
)
```

**Reply by bluebill1049** [reply]
hey @pranavbadami how could your policy worked without even a table join (your above SQL)? I mean `team` and `app_users`. In fact, I am having difficulties applying a similar policy like above, as soon as start joining the tables, the policy fails.  https://github.com/supabase/supabase/discussions/2978

**Reply by pranavbadami** [reply]
Hey @bluebill1049 , could you provide a bit more info on your situation? Basically, I had a linking value in each table and my `app_users` table that linked everything to the same `team` value. I was then able to avoid having to join across tables in my RLS policies.

**Reply by bluebill1049** [reply]
thanks, @pranavbadami for the reply. I left the question inside your query. Relative new to SQL, correct me if i am wrong please.

```sql
EXISTS (
   SELECT app_users.slack_id,
    app_users.name
   FROM app_users
  WHERE app_users.team = teams.slack_id // my question is on this part, how could temas.slack_id aware of the relationship without the join, does that means you have a foreign key set up with the app_users table so you don't need to join the tables?
  AND app_users.auth_users_uid = uid()
)
```

**Comment by thinusn** [comment]
How would the row level security policy on the `app_users` table look like so that the authenticated Supabase auth.user can only access the rows in `app_users` for teams that they are a part of i.e. only members of a team can view existing and add new members to that team ? 

Continuing from the example above: When selecting from `app_users` for user `example123` it should return all members from team1 and all members from team2 (since they are a member of both).  When inserting a new entry into the `app_users` table for team1 only `example123` can add members and  conversely for `team2`  both `example123` and `example456` can add/remove new users to `team2`. I played around with it a bit but kept getting `infinite recursion detected in policy for relation` responses when querying `app_users`

**Reply by pranavbadami** [reply]
Hey @thinusn , I had the same situation as you when I was trying to use a JOIN and realized what I needed was much simpler. Here's what ended up working for me as a policy on the `app_users` table: 

```
EXISTS (
   SELECT app_users.slack_id,
    app_users.name
   FROM app_users
  WHERE app_users.team = teams.slack_id 
  AND app_users.auth_users_uid = uid()
)
```
Basically, you're checking if an entry exists in the `users` table with the same `uid` as the accessing user and the same `team` as the given row you are accessing in `app_users`.

**Reply by thinusn** [reply]
Regarding `app_users.team = teams.slack_id` how did you define `users.team` if this policy is on `app_users` ? Does it mean the relationship is maintained in two places?

**Reply by pranavbadami** [reply]
Hey, so I think I misunderstood your question. The issue I was discussing here was for accessing rows on the `teams` table, not the `app_users` table. I also fixed the example above which was confusing because I said `users.team` where I really meant `app_users.team`. 

I understand your question now, you're asking how do you see all the entries in `app_users` that belong to the same `team` you're in. I'm assuming you tried something like this but then had the recursion issue (which makes sense):

```
EXISTS ( 
  SELECT
    au.slack_id,
    au.name
   FROM app_users AS au
  WHERE au.team = team <--- pseudocoded here so aliasing might be off
  AND au.auth_users_uid = uid()
)
```

Unfortunately, I don't have a solution but this is something I'll probably need to figure out as well. One initial approach could be the one detailed in this [thread](https://www.postgresql-archive.org/Recursive-row-level-security-policy-td5935014.html).

**Reply by steve-chavez** [reply]
> One initial approach could be the one detailed in this thread.

I can vouch for the security definer function workaround for clearing infinite recursion in policies. I've done this in my example here: [policy](https://github.com/steve-chavez/socnet/blob/master/security/users_blocked.sql#L113), [sdf](https://github.com/steve-chavez/socnet/blob/master/schema/util.sql#L59-L66).

**Reply by thinusn** [reply]
Thanks for the help @pranavbadami and @steve-chavez. The following worked for me (adapted for example from above):
```
create or replace function get_teams_for_user(user_id uuid) returns setof bigint as $$
  select team_id from app_users where user_id = $1
$$ stable language sql security definer;

create policy "Allow individual read access if member of a team" on app_users
for select using (
  team_id in ( select get_teams_for_user(auth.uid()))
);
```

:)

**Reply by pranavbadami** [reply]
Awesome, thanks for sharing, this will be helpful for me moving forward too! 🙌

**Reply by kennethcassel** [reply]
Why does moving the `select team_id from app_users where user_id = $1` query into a function make the policy work? We saw this in our own codebase as well for a similar issue.

We were getting a recursive issue on the policy until we added a function for the query logic.

**Reply by b2m9** [reply]
Sorry to dig out an old thread, but I would like to echo @kennethcassel's question: why does using a function solve the recursion issue?

I understand that functions are bypassing RLS. But the functions themselves are callable from the client, which isn't ideal. Or do I understand this wrongly?

**Reply by steve-chavez** [reply]
> why does using a function solve the recursion issue?

Because the security definer function breaks out of the recursive loop. Security definer: the user that created the function, which doesn't evaluate the policy(since it bypasses RLS).

> But the functions themselves are callable from the client, which isn't ideal. Or do I understand this wrongly?

In the example I posted [above](https://github.com/supabase/supabase/discussions/811#discussioncomment-558183), the SDF(security definer function) is on an `util` schema, which is non-callable by the client.

You could `create schema private` and define the function as `create function private.my_func` to make this more explicit. Remember that only db objects inside the `public` schema are callable by the client.

**Reply by b2m9** [reply]
> You could create schema private and define the function as create function private.my_func to make this more explicit.

This would be so helpful to have explicitly spelled out in the Supabase admin interface. Thank you so much, I wouldn't have known that this is even possible. It really bugged me that the SDF was callable as an RPC on the client.

**Comment by scomma** [comment]
> has a column auth_users_id that is a reference to auth.users.uid (side question: is it possible to foreign key here? I didn't find a way to via the console)

Just want to add that this is definitely possible but you need to write the SQL yourself. Mine looks like this:

```
create table public.accounts (
  id uuid references auth.users not null,
  ...
  primary key (id)
);
```
